src/Storages/S3Queue/S3QueueSource.cpp

```diff
@@ -218,61 +118,86 @@ String StorageS3QueueSource::getName() const
 
 Chunk StorageS3QueueSource::generate()
 {
-    auto file_progress = getContext()->getFileProgressCallback();
     while (true)
     {
-        if (isCancelled() || !reader)
+        if (!reader)
+            break;
+
+        if (isCancelled())
         {
-            if (reader)
-                reader->cancel();
+            reader->cancel();
+            break;
+        }
+
+        if (shutdown_called)
+        {
+            if (processed_rows_from_file)
+            {
+                /// We could delay shutdown until files, which already started processing before the shutdown, finished.
+                /// But if files are big and `s3queue_processing_threads_num` is not small, it can take a significant time.
+                /// Anyway we cannot do anything in case of SIGTERM, so destination table must anyway support deduplication,
+                /// so here we will rely on it here as well.
```

Comment:
Also in case of exception during parsing we could already insert some data into destination table and after retries there will be duplications. We should defenitely add a note about it in the documentation. And I don't think we can solve it by saving file offsets that we processed because of formats with metadata and random access like Parquet/ORC/Arrow. Maybe we can store processed rows for each file and when we start reading the file again, we can skip already processed rows by just ignoring them here after `reader->pull()` (also in future we can add optimization for it and add method skipRows(size_t rows) for input formats because most formats can skip rows fast enough). Or combine both methods (saving offset and saving processed rows) and use one of them depending on format (like, for CSV/TSV/JSONEachRow/etc we can save offset and start reading from it, and for Parquet/ORC/etc we can save previrously read rows and skip them before returning blocks (and for such formats skipping rows can be optimized using their metadata).

Reply:
> Maybe we can store processed rows for each file and when we start reading the file again, we can skip already processed rows by just ignoring them here after reader->pull() (also in future we can add optimization for it and add method skipRows(size_t rows) for input formats because most formats can skip rows fast enough). 

Good idea. Though if we update processed rows count in keeper for every block, it will result in too many keeper requests...

> We should defenitely add a note about it in the documentation. 

Will do 👌🏻 .